File: 0009/problem9.c
#include <stdio.h>
#include <math.h>

int check(int a, int n) {
    //Check if a given a and n pair can work for a^2+b^2=c^2 where a+b+c=n
    float a_f = (float) a;
    float n_f = (float) n;
    float b = (n_f/2-a_f)/(1-a_f/n_f);
    if(b == (int)b) {
        int c = n-a-b;
        if(a<b && b<c) {
            return a*b*c;
        }
    }
    return 0;
}

int special_pythagorean_triplet(int n) {
    //Find all the integer factors of n
    int soln;
    for (int a = 1; a <= sqrt(n); a++) {
        if (n%a == 0) {
            if (n/a == a) {
                if(soln = check(a, n))
                    return soln;
            }
            else {
                if(soln = check(n/a, n))
                    return soln;
            }
        }
    }
}

int main() {
    printf("%d\n", special_pythagorean_triplet(1000));
}

File: 0009/problem9.py
from math import sqrt

def factors_of(n):
    """Find all the integer factors of n"""
    for i in range(2, int(sqrt(n))):
        if n%i == 0:
            yield(i)
            if i != n/i:
                yield(n/i)

def special_pythagorean_triplet(n):
    """Find the pythagorean triplet that sums to n and return it's product"""
    for a in factors_of(n):
        # For each factor find b such that a^2+b^2 = (1000-a-b)^2
        b = (n/2-a)/(1-a/n)
        # For the a and b find c
        c = n-a-b
        # Check if the triplet is valid
        if a<b<c and b==int(b):
            return(int(a*b*c))

print(special_pythagorean_triplet(1000))